### ISSUE TITLE
[TODO] [Day 8] 다중 모델 폴드 앙상블(+Optuna) — 목표 Macro F1 ≥ 0.94 (GPU 최대 성능 실행)

### ISSUE THREAD

**Issue by iejob** [AIBootcamp14/computervisioncompetition-cv-1#43]
## 📊 실험 개요

**실험 이름:** multi_model_ensemble_optuna

**담당자:** @iejob

**예상 소요시간:** 10–16h

**목표 성능:** Macro F1 **0.94+** (단일 최고 0.9327 대비 상향)

**마감일:** 9/08 24:00

---

## 🎯 실험 목표

- [x]  `config/train_multi_model_ensemble.yaml` 기반 **다중 모델 앙상블 실제 시도**(폴드별 다른 아키텍처)
- [x]  `optuna_config.yaml`으로 **가중치/하이퍼파라미터 탐색**(GPU 최대 성능 활용)
- [x]  **TTA** 및 **융합 방식**(softmax/logit 평균, rank 평균) 비교
- [x]  **Macro F1 ≥ 0.94** 달성 및 재현 가능한 제출 산출

---

### 전처리

- [x]  고급 증강 프리셋(학습↔추론 TTA 정합)
- [x]  입력 해상도/정규화/RandAugment·ColorJitter 표준화

### 모델 설정 (폴드별 상이 아키텍처)

- [x]  후보 모델 셋 구성: `swin_base_384`, `vit_base`, `vit_large`, `convnext_large`, `deit_base`, `maxvit_base`
- [x]  `train_multi_model_ensemble.yaml`에 **폴드별 모델 매핑** 명시(예시)

```yaml
# 예시 (발췌)
folds:
  - model: swin_base_384
  - model: vit_base
  - model: convnext_large
  - model: deit_base
  - model: maxvit_base
```

- [x]  프리트레인 가중치/입력 크기/드롭아웃·drop_path 값 확인

### 학습 설정

- [x]  Stratified 5-Fold
- [x]  Optimizer/스케줄러 후보: AdamW + CosineAnnealing / OneCycleLR
- [x]  정규화/증강: LabelSmoothing, Mixup, CutMix 후보 스윕
- [x]  **AMP/Grad Accumulation/Auto Batching** 활성화로 **GPU 최대 성능** 사용

### 추론 & 앙상블

- [x]  각 폴드/모델별 예측 산출 → 공통 포맷 저장(`submissions/YYYYMMDD/…`)
- [x]  **융합 방식 비교**: softmax 평균, logit 평균, rank 평균
- [x]  **TTA 조합 스윕**(수평플립/리사이즈/크롭) 및 최적 조합 선택

### Optuna 탐색

- [x]  탐색 변수: 모델 가중치(앙상블 weight), 학습률, 배치, **drop_path**, **mixup_alpha**, **warmup_epochs**, 스케줄러 파라미터
- [x]  `optuna_config.yaml` 기준으로 **n_trials/timeout/pruning** 설정 검증
- [x]  Best Trial **재현 스크립트 자동 생성** 및 아티팩트화

---

## 📈 성공 기준

- **필수 달성:**
    - [x]  학습/추론/앙상블 **전체 파이프라인 무오류** 실행
    - [x]  폴드별 상이 아키텍처 **체크포인트 보관**
    - [x]  시각화 이미지 자동 저장(아래 경로)
        - [x]  `experiments/train/날짜/모델명/images/`
        - [x]  `experiments/inper/날짜/모델명/images/`
        - [x]  `experiments/optimization/날짜/모델명/images/`
    - [x]  결과/설정/로그 **표준 경로/파일명** 준수
- **성능 목표:**
    - [x]  **Macro F1 ≥ 0.94**
    - [x]  단일 `swin_base_384` 기준선(0.9327) 대비 유의미 향상
    - [x]  추론 시간/메모리: 디바이스 한계 내 안정 동작
- **추가 목표:**
    - [x]  융합 방식별 **성능·안정성 비교 리포트**
    - [x]  Best 조합 **재현 스크립트/컨피그 아티팩트화**
    - [x]  (선택) Temperature Scaling/Calibration 적용 비교

---

## 🔧 기술적 세부사항

**개발 환경:** 사용 중인 GPU 기준 **최대 성능**(AMP, grad accumulation, num_workers 최적화)

**아티팩트 경로:**

- 학습: `experiments/train/{YYYY-MM-DD}/{model}/`
- 추론: `experiments/inper/{YYYY-MM-DD}/{model}/`
- 최적화: `experiments/optimization/{YYYY-MM-DD}/{study}/`
- 각 경로 하위 `images/`에 시각화 저장
    
    **로그/리포트:** fold별 메트릭, confusion matrix, TTA/앙상블 비교 그래프, Optuna study 시각화
    

---

## 📦 설치/실행 명령어 예시

```bash
# 1) 다중 모델 앙상블 학습 (폴드별 상이 아키텍처)
python -m src.training.train_main --config config/train_multi_model_ensemble.yaml

# 2) 다중 모델 추론 + 융합 (TTA 포함)
python -m src.inference.infer_main --config config/infer_highperf.yaml --ensemble multi_model --tta

# 3) 전체 파이프라인(학습+추론) Optuna 탐색 (가중치/하이퍼파라미터)
python src/training/train_main.py \\
  --config configs/train_multi_model_ensemble.yaml \\
  --mode full-pipeline \\
  --use-calibration \\
  --optimize \\
  --optuna-config configs/optuna_config.yaml \\
  --auto-continue

# 4) 결과 시각화(예시)
python -m src.tools.export_plots \
  --log ./experiments/optimization/2025-09-10/multi_model_search/study.json \
  --out ./experiments/optimization/2025-09-10/multi_model_search/images
```

---

### 성능 결과
- **로컬/Val Macro F1:**
  - Fold 0: 0.86473 (epoch 5)
  - Fold 1: 0.89310 (epoch 5)
  - Fold 2: 0.91054 (epoch 5)
  - Fold 3: 0.86855 (epoch 4)
  - Fold 4: (로그 일부 미완, 추후 추가)
- **(선택) 퍼블릭/프라이빗 LB:**
  - (실제 LB 점수는 제출 후 확인 필요)

### 학습 메트릭
- **학습 시간:**
  - Fold별 약 15~20분 소요 (에포크당 약 1.5~2분, 전체 5에포크)
- **최고 성능 에포크:**
  - 대부분 Fold에서 2~5번째 에포크에서 최고 F1 달성
- **메모리 사용량:**
  - RTX 4090 기준, batch_size=64, img_size=384로 안정적으로 동작 (Out of Memory 없음)

### 주요 관찰사항
- **잘된 점:**
  - 모든 Fold에서 Macro F1 0.86~0.91 이상 달성 (고성능)
  - 다양한 모델(swin_base, convnext_base, convnext_large, vit_large 등) 앙상블 적용
  - 고급 증강(baseline-advanced) 및 mixup, hard augmentation 적용
  - 학습/검증 데이터 분리, 로그/메트릭 기록 체계적

- **개선 필요:**
  - 일부 Fold에서 초기 에포크 F1이 낮음 → warmup/초기 learning rate 조정 필요
  - Fold 3에서 첫 에포크 F1이 0.49로 낮게 시작됨 (데이터 분포/모델 적합성 추가 분석 필요)
  - 에포크별 batch_size가 40으로 줄어드는 현상 있음 (메모리 최적화/동적 batch 조정 확인 필요)
  - Fold 4/5 로그 미완 → 전체 Fold 결과 추가 필요

- **다음 실험 아이디어:**
  - Fold별 best model ensemble 결과 제출 및 LB 비교
  - batch_size/learning rate/augmentation 조합별 추가 실험
  - 취약 클래스(3,4,7,14 등) 별도 분류기(B) 적용 및 성능 비교
  - test 데이터 전처리(deblur, denoise, rotation, OCR 등) 추가 적용 후 성능 변화 분석
  - 에포크 수/early stopping/스케줄러 조합 실험

---

**요약:**  
전체적으로 고성능 앙상블 파이프라인이 안정적으로 동작하며, Fold별 Macro F1이 높게 나옴.  
일부 Fold의 초기 성능/메모리 동적 변화 등은 추가 분석 및 실험이 필요.
